### PR TITLE
Fix: add type check for struct before downcasting

### PR DESCRIPTION
**Pull request by HarshitaKalani**
Towards: #3732

### CODE REVIEW

File: src/libasr/codegen/asr_to_llvm.cpp [hunk 1]
```diff
@@ -8696,11 +8698,15 @@ class ASRToLLVMVisitor : public ASR::BaseVisitor<ASRToLLVMVisitor>
                     builder->CreateBitCast(dt, llvm::Type::getVoidTy(context)->getPointerTo()),
                     polymorphic_addr);
                 llvm::Value* type_id_addr = llvm_utils->create_gep(abstract_, 0);
-                ASR::Struct_t* struct_t = ASR::down_cast<ASR::Struct_t>(arg_type);
```
Comment (HarshitaKalani) [inline src/libasr/codegen/asr_to_llvm.cpp:8699]: The problem is that here the arg_type is Character_t. How to handle that? And is this the right place to make the changes?

File: b.f90 [hunk 2]
```diff
@@ -0,0 +1,10 @@
+program main
+    call p(1)
+    contains
+
+    subroutine p(a)
+        Class(*), intent(in) :: a
+        print *, "Hello World"
+    end subroutine
+
+end program main
```
Comment (HarshitaKalani) [inline b.f90:10]: This example started working with the changes here

File: a.f90 [hunk 3]
```diff
@@ -0,0 +1,23 @@
...
+    ! This should not compile
+    !          call f(generic)
+        end subroutine print_generic
+    
+        subroutine f(x)
+        character(*), intent(in) :: x
+        print *, x
+        end subroutine
+    
+end
```
Comment (HarshitaKalani) [inline a.f90:23]: This gives:
```console
 lfortran a.f90
; ModuleID = 'LFortran'
source_filename = "LFortran"

%"~abstract_type_polymorphic" = type { i64, void* }
%"__vtab_~abstract_type" = type { i64 }

@0 = private unnamed_addr constant [2 x i8] c" \00", align 1
@1 = private unnamed_addr constant [2 x i8] c"\0A\00", align 1
@2 = private unnamed_addr constant [5 x i8] c"%s%s\00", align 1
@3 = private unnamed_addr constant [4 x i8] c"abc\00", align 1

define void @f(i8** %x) {
.entry:
  %0 = load i8*, i8** %x, align 8
  call void (i8*, ...) @_lfortran_printf(i8* getelementptr inbounds ([5 x i8], [5 x i8]* @2, i32 0, i32 0), i8* %0, i8* getelementptr inbounds ([2 x i8], [2 x i8]* @1, i32 0, i32 0))
  br label %return

return:                                           ; preds = %.entry
  ret void
}

define void @print_generic(%"~abstract_type_polymorphic"* %generic, i1* %is_generic_present_) {
.entry:
  %0 = alloca %"__vtab_~abstract_type", align 8
  %1 = getelementptr %"__vtab_~abstract_type", %"__vtab_~abstract_type"* %0, i32 0, i32 0
  store i64 0, i64* %1, align 4
  %2 = getelementptr %"~abstract_type_polymorphic", %"~abstract_type_polymorphic"* %generic, i32 0, i32 0
  %3 = load i64, i64* %2, align 4
  %4 = icmp eq i64 %3, -5
  br i1 %4, label %then, label %else

then:                                             ; preds = %.entry
  call void @f(%"~abstract_type_polymorphic"* %generic)
  br label %ifcont

else:                                             ; preds = %.entry
  br label %ifcont

ifcont:                                           ; preds = %else, %then
  br label %return

return:                                           ; preds = %ifcont
  ret void
}

declare void @_lfortran_printf(i8*, ...)

define i32 @main(i32 %0, i8** %1) {
.entry:
  %call_arg_value1 = alloca i1, align 1
  %call_arg_value = alloca i8*, align 8
  call void @_lpython_call_initial_functions(i32 %0, i8** %1)
  store i8* getelementptr inbounds ([4 x i8], [4 x i8]* @3, i32 0, i32 0), i8** %call_arg_value, align 8
  %2 = alloca %"~abstract_type_polymorphic", align 8
  %3 = getelementptr %"~abstract_type_polymorphic", %"~abstract_type_polymorphic"* %2, i32 0, i32 1
  %4 = bitcast i8** %call_arg_value to void*
  store void* %4, void** %3, align 8
  %5 = getelementptr %"~abstract_type_polymorphic", %"~abstract_type_polymorphic"* %2, i32 0, i32 0
  store i1 true, i1* %call_arg_value1, align 1
  call void @print_generic(%"~abstract_type_polymorphic"* %2, i1* %call_arg_value1)
  ret i32 0
}

declare void @_lpython_call_initial_functions(i32, i8**)
asr_to_llvm: module failed verification. Error:
Call parameter type does not match function signature!
%"~abstract_type_polymorphic"* %generic
 i8**  call void @f(%"~abstract_type_polymorphic"* %generic)

code generation error: asr_to_llvm: module failed verification. Error:
Call parameter type does not match function signature!
%"~abstract_type_polymorphic"* %generic
 i8**  call void @f(%"~abstract_type_polymorphic"* %generic)



Note: Please report unclear, confusing or incorrect messages as bugs at
https://github.com/lfortran/lfortran/issues.
```

**Review by certik — APPROVED** [review]
I think that's fine for now.

**Pull request merged**